File: app/src/main/java/com/aexon/core/AexonMath.java
/*
* Copyright (c) 2026 Fora
* 
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <https://www.gnu.org/licenses/>.
* 
* Contact: Fora <fora060823@gmail.com>
* Created: 27-01-2026
*/
package com.aexon.core.core;

import android.content.Context;
import android.content.res.Resources;
import android.os.Build;
import android.util.TypedValue;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

@RequiresApi(api = Build.VERSION_CODES.O)
public final class AexonMath {

    private AexonMath() {
        throw new UnsupportedOperationException("No instances");
    }

    public static float dpToPx(@NonNull Context context, float dp) {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.getResources().getDisplayMetrics());
    }

    public static float dpToPx(float dp) {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, Resources.getSystem().getDisplayMetrics());
    }

    public static int dpToPxInt(@NonNull Context context, float dp) {
        return (int) dpToPx(context, dp);
    }

    public static int dpToPxInt(float dp) {
        return (int) dpToPx(dp);
    }

    public static float pxToDp(@NonNull Context context, float px) {
        return px / context.getResources().getDisplayMetrics().density;
    }

    public static float pxToDp(float px) {
        return px / Resources.getSystem().getDisplayMetrics().density;
    }

    public static float spToPx(@NonNull Context context, float sp) {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, context.getResources().getDisplayMetrics());
    }

    public static float spToPx(float sp) {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, Resources.getSystem().getDisplayMetrics());
    }

    public static int spToPxInt(@NonNull Context context, float sp) {
        return (int) spToPx(context, sp);
    }

    public static int spToPxInt(float sp) {
        return (int) spToPx(sp);
    }

    public static float pxToSp(@NonNull Context context, float px) {
        return px / context.getResources().getDisplayMetrics().scaledDensity;
    }

    public static float pxToSp(float px) {
        return px / Resources.getSystem().getDisplayMetrics().scaledDensity;
    }

    public static float clamp(float value, float minValue, float maxValue) {
        return Math.max(minValue, Math.min(maxValue, value));
    }

    public static int clamp(int value, int minValue, int maxValue) {
        return Math.max(minValue, Math.min(maxValue, value));
    }

    public static float clamp01(float value) {
        return clamp(value, 0f, 1f);
    }

    public static float lerp(float start, float end, float fraction) {
        return start + fraction * (end - start);
    }

    public static int lerp(int start, int end, float fraction) {
        return Math.round(start + fraction * (end - start));
    }

    public static float inverseLerp(float start, float end, float value) {
        if (start == end) return 0f;
        return clamp01((value - start) / (end - start));
    }

    public static float map(float value, float inMin, float inMax, float outMin, float outMax) {
        float t = inverseLerp(inMin, inMax, value);
        return lerp(outMin, outMax, t);
    }

    public static float roundTo(float value, int decimals) {
        float factor = (float) Math.pow(10.0, decimals);
        return Math.round(value * factor) / factor;
    }

    public static float distance(float x1, float y1, float x2, float y2) {
        float dx = x2 - x1;
        float dy = y2 - y1;
        return (float) Math.sqrt(dx * dx + dy * dy);
    }

    public static boolean isApprox(float a, float b) {
        return isApprox(a, b, 0.0001f);
    }

    public static boolean isApprox(float a, float b, float epsilon) {
        return Math.abs(a - b) < epsilon;
    }

    public static int wrap(int value, int min, int max) {
        int range = max - min;
        if (range <= 0) return min;
        int result = (value - min) % range;
        if (result < 0) result += range;
        return result + min;
    }
}

File: app/src/main/java/com/aexon/core/AexonMath.kt
package com.aexon.core

import android.content.Context
import android.content.res.Resources
import android.util.TypedValue
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

class AexonMath {
	companion object {
		
		@JvmStatic
		fun dpToPx(context: Context, dp: Float): Float {
			return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.resources.displayMetrics)
		}
		
		@JvmStatic
		fun dpToPx(dp: Float): Float {
			return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, Resources.getSystem().displayMetrics)
		}
		
		@JvmStatic
		fun dpToPxInt(context: Context, dp: Float): Int {
			return dpToPx(context, dp).toInt()
		}
		
		@JvmStatic
		fun dpToPxInt(dp: Float): Int {
			return dpToPx(dp).toInt()
		}
		
		@JvmStatic
		fun pxToDp(context: Context, px: Float): Float {
			return px / context.resources.displayMetrics.density
		}
		
		@JvmStatic
		fun pxToDp(px: Float): Float {
			return px / Resources.getSystem().displayMetrics.density
		}
		
		@JvmStatic
		fun spToPx(context: Context, sp: Float): Float {
			return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, context.resources.displayMetrics)
		}
		
		@JvmStatic
		fun spToPx(sp: Float): Float {
			return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, Resources.getSystem().displayMetrics)
		}
		
		@JvmStatic
		fun spToPxInt(context: Context, sp: Float): Int {
			return spToPx(context, sp).toInt()
		}
		
		@JvmStatic
		fun spToPxInt(sp: Float): Int {
			return spToPx(sp).toInt()
		}
		
		@JvmStatic
		fun pxToSp(context: Context, px: Float): Float {
			return px / context.resources.displayMetrics.scaledDensity
		}
		
		@JvmStatic
		fun pxToSp(px: Float): Float {
			return px / Resources.getSystem().displayMetrics.scaledDensity
		}
		
		@JvmStatic
		fun clamp(value: Float, minValue: Float, maxValue: Float): Float {
			return max(minValue, min(maxValue, value))
		}
		
		@JvmStatic
		fun clamp(value: Int, minValue: Int, maxValue: Int): Int {
			return max(minValue, min(maxValue, value))
		}
		
		@JvmStatic
		fun clamp01(value: Float): Float {
			return clamp(value, 0f, 1f)
		}
		
		@JvmStatic
		fun lerp(start: Float, end: Float, fraction: Float): Float {
			return start + fraction * (end - start)
		}
		
		@JvmStatic
		fun lerp(start: Int, end: Int, fraction: Float): Int {
			return (start + fraction * (end - start)).roundToInt()
		}
		
		@JvmStatic
		fun inverseLerp(start: Float, end: Float, value: Float): Float {
			if (start == end) return 0f
			return clamp01((value - start) / (end - start))
		}
		
		@JvmStatic
		fun map(value: Float, inMin: Float, inMax: Float, outMin: Float, outMax: Float): Float {
			val t = inverseLerp(inMin, inMax, value)
			return lerp(outMin, outMax, t)
		}
		
		@JvmStatic
		fun roundTo(value: Float, decimals: Int): Float {
			val factor = Math.pow(10.0, decimals.toDouble()).toFloat()
			return (value * factor).roundToInt() / factor
		}
		
		@JvmStatic
		fun distance(x1: Float, y1: Float, x2: Float, y2: Float): Float {
			val dx = x2 - x1
			val dy = y2 - y1
			return sqrt(dx * dx + dy * dy)
		}
		
		@JvmStatic
		fun isApprox(a: Float, b: Float, epsilon: Float = 0.0001f): Boolean {
			return abs(a - b) < epsilon
		}
		
		@JvmStatic
		fun wrap(value: Int, min: Int, max: Int): Int {
			val range = max - min
			if (range <= 0) return min
			var result = (value - min) % range
			if (result < 0) result += range
			return result + min
		}
	}
}